From 0df0303781e9d7c59bd51935c618c967b249e95c Mon Sep 17 00:00:00 2001
From: Serguei Katkov <serguei.i.katkov@intel.com>
Date: Thu, 22 Nov 2012 13:55:16 +0700
Subject: Dalvik: a fix in Interpreter for wrong location for spilling ebx

BZ: 75820

Interpreter's common_selectTrace and common_selectTrace2 uses the same memory location
for spilling ebx as Interpreter's entry point. As a result ebx spilled in Interpreter's
entry point is overriden and Interpreter restores invalid value of ebx.
The patch changes the spilling location for common_selectTrace and common_selectTrace2 to
SPILL_TMP1.

Category: device-enablement
Domain: AOSP-Dalvik-Interpreter
Origin: internal
Upstream-Candidate: yes

Change-Id: Ib0d481637270883641061c5f3ec2219437de13e2
Orig-MCG-Change-Id: I8a0f5451bf3bfa0e1ade63ed68611ca2ee76cb29
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/mterp/out/InterpAsm-x86.S |    8 ++++----
 vm/mterp/x86/footer.S        |    8 ++++----
 2 files changed, 8 insertions(+), 8 deletions(-)

diff --git a/vm/mterp/out/InterpAsm-x86.S b/vm/mterp/out/InterpAsm-x86.S
index 2081fcc..057f38c 100644
--- a/vm/mterp/out/InterpAsm-x86.S
+++ b/vm/mterp/out/InterpAsm-x86.S
@@ -15976,7 +15976,7 @@ common_Profile:
     movl   $kJitTSelectRequest,%eax
     # On entry, eax<- jitState, rPC valid
 common_selectTrace:
-    mov         %ebx, EBX_SPILL(%ebp)
+    SPILL_TMP1(%ebx)
     movl        rSELF, %ebx
     movzwl      offThread_subMode(%ebx), %ecx
     and         $(kSubModeJitTraceBuild | kSubModeJitSV), %ecx
@@ -15995,7 +15995,7 @@ common_selectTrace:
    SAVE_PC_FP_TO_SELF %ecx
    call dvmJitCheckTraceRequest
 3:
-   mov          EBX_SPILL(%ebp), %ebx
+   UNSPILL_TMP1(%ebx)
    FETCH_INST
    movl rSELF, %ecx
    movl offThread_curHandlerTable(%ecx), rIBASE
@@ -16003,12 +16003,12 @@ common_selectTrace:
    GOTO_NEXT
 
 common_selectTrace2:
-    mov         %ebx, EBX_SPILL(%ebp)
+    SPILL_TMP1(%ebx)
     movl        rSELF, %ebx
     movl        %ebx, OUT_ARG0(%esp)
     movb        %al, offThread_jitState(%ebx)
     movzwl      offThread_subMode(%ebx), %ecx
-    mov         EBX_SPILL(%ebp), %ebx
+    UNSPILL_TMP1(%ebx)
     and         (kSubModeJitTraceBuild | kSubModeJitSV), %ecx
     jne         3f                     # already doing JIT work, continue
 
diff --git a/vm/mterp/x86/footer.S b/vm/mterp/x86/footer.S
index bc05e3b..dfb3d68 100644
--- a/vm/mterp/x86/footer.S
+++ b/vm/mterp/x86/footer.S
@@ -299,7 +299,7 @@ common_Profile:
     movl   $$kJitTSelectRequest,%eax
     # On entry, eax<- jitState, rPC valid
 common_selectTrace:
-    mov         %ebx, EBX_SPILL(%ebp)
+    SPILL_TMP1(%ebx)
     movl        rSELF, %ebx
     movzwl      offThread_subMode(%ebx), %ecx
     and         $$(kSubModeJitTraceBuild | kSubModeJitSV), %ecx
@@ -318,7 +318,7 @@ common_selectTrace:
    SAVE_PC_FP_TO_SELF %ecx
    call dvmJitCheckTraceRequest
 3:
-   mov          EBX_SPILL(%ebp), %ebx
+   UNSPILL_TMP1(%ebx)
    FETCH_INST
    movl rSELF, %ecx
    movl offThread_curHandlerTable(%ecx), rIBASE
@@ -326,12 +326,12 @@ common_selectTrace:
    GOTO_NEXT
 
 common_selectTrace2:
-    mov         %ebx, EBX_SPILL(%ebp)
+    SPILL_TMP1(%ebx)
     movl        rSELF, %ebx
     movl        %ebx, OUT_ARG0(%esp)
     movb        %al, offThread_jitState(%ebx)
     movzwl      offThread_subMode(%ebx), %ecx
-    mov         EBX_SPILL(%ebp), %ebx
+    UNSPILL_TMP1(%ebx)
     and         (kSubModeJitTraceBuild | kSubModeJitSV), %ecx
     jne         3f                     # already doing JIT work, continue
 
-- 
1.7.4.1

